### PR TITLE
Improve performance when logging is disabled

### PR DESCRIPTION
**Pull request by tomer-w**
Improve the logging performance by not running expensive JSON.stringify when the logging is disabled.
The old pattern was something like that:
```
logger.debug(`sendZclFrameToAll ${JSON.stringify(payload)}`, NS);
```
This was calling the ```stringify``` function even if eventually the logger was disabled.
The new pattern is:
```
logger.debug(() => `sendZclFrameToAll ${JSON.stringify(payload)}`, NS);
```
Which will not call the lambda if the tracing is disabled.
The logger function will accept both string messages as they used to be and the new lambda format, so there is no need to change all logging calls, only the ones that are deemed expensive.

I am also adding ```Logger.isEnabled``` function which can be used by the caller to decide if there is a need to prepare complex log message.

I also changed all the expensive calls to the logger in the zigbee-herdsman to use the new functionality:
https://github.com/Koenkk/zigbee-herdsman/pull/1178

### CODE REVIEW

File: lib/util/logger.ts [hunk 1]
```diff
@@ -189,32 +189,40 @@ class Logger {
         return this.cachedNamespacedLevels[namespace];
     }
 
-    private log(level: settings.LogLevel, message: string, namespace: string): void {
+    private log(level: settings.LogLevel, messageOrLambda: string | (() => string), namespace: string): void {
+        if (this.isEnabled(level, namespace)) {
+            const message: string = typeof messageOrLambda === 'function' ? messageOrLambda() : messageOrLambda;
```
Comment (Nerivec) [inline lib/util/logger.ts:194]: Should be able to do `messageOrLambda instanceof Function` here (avoid string compare, should be a bit faster).

File: lib/util/logger.ts [hunk 2]
```diff
@@ -189,32 +189,40 @@ class Logger {
...
+            this.logger.log(level, `${namespace}: ${message}`);
+        }
+    }
+
+    public isEnabled(level: settings.LogLevel, namespace: string): boolean {
```
Comment (Nerivec) [inline lib/util/logger.ts:199]: Don't think this is needed, the entire codebase is unaware of log level (except logger stuff ofc), probably best it stays that way.
I'd say best to avoid another function call in `log` and leave this as before.

**Comment by Nerivec** [comment]
Should provide a decent performance boost:

```log
┌─────────┬────────────────────────────────┬─────────────┬────────────────────┬──────────┬──────────┐
│ (index) │ Task Name                      │ ops/sec     │ Average Time (ns)  │ Margin   │ Samples  │
├─────────┼────────────────────────────────┼─────────────┼────────────────────┼──────────┼──────────┤
│ 0       │ 'plain log'                    │ '275,104'   │ 3634.9870738169716 │ '±1.29%' │ 1375521  │
│ 1       │ 'plain log new'                │ '273,253'   │ 3659.602662925059  │ '±1.41%' │ 1366291  │
│ 2       │ 'stringify log'                │ '226,876'   │ 4407.674646871684  │ '±0.79%' │ 1134385  │
│ 3       │ 'stringify log new'            │ '226,783'   │ 4409.496455342999  │ '±0.82%' │ 1133917  │
│ 4       │ 'plain log - disabled'         │ '7,542,096' │ 132.5891300310219  │ '±0.15%' │ 37710482 │
│ 5       │ 'plain log new - disabled'     │ '7,154,958' │ 139.7632147654249  │ '±0.47%' │ 35774793 │
│ 6       │ 'stringify log - disabled'     │ '1,706,746' │ 585.9101057924765  │ '±0.15%' │ 8533733  │
│ 7       │ 'stringify log new - disabled' │ '5,587,460' │ 178.9721728850612  │ '±1.57%' │ 27937304 │
└─────────┴────────────────────────────────┴─────────────┴────────────────────┴──────────┴──────────┘
```

<details><summary>code</summary>

```typescript
// https://github.com/tinylibs/tinybench
import {Bench, hrtimeNow} from 'tinybench';
import logger from './logger.js';
import loggernew from './loggernew.js';

const bench = new Bench({time: 5000, warmupTime: 1000, now: hrtimeNow});

logger.init(); // set to 'info' with silenced Console transport
loggernew.init(); // set to 'info' with silenced Console transport

bench.add('plain log', function() {
    logger.error('test');
});

bench.add('plain log new', function() {
    loggernew.error('test');
});

bench.add('stringify log', function() {
    logger.error(`test ${JSON.stringify({ a: 1, b: 2 })}`);
});

bench.add('stringify log new', function() {
    loggernew.error(() => `test ${JSON.stringify({ a: 1, b: 2 })}`);
});

bench.add('plain log - disabled', function() {
    logger.debug('test');
});

bench.add('plain log new - disabled', function() {
    loggernew.debug('test');
});

bench.add('stringify log - disabled', function() {
    logger.debug(`test ${JSON.stringify({ a: 1, b: 2 })}`);
});

bench.add('stringify log new - disabled', function() {
    loggernew.debug(() => `test ${JSON.stringify({ a: 1, b: 2 })}`);
});

await bench.warmup(); // make results more reliable, ref: https://github.com/tinylibs/tinybench/pull/50
await bench.run();

console.table(bench.table());

await logger.end();
await loggernew.end();
```

</details>

**Comment by tomer-w** [comment]
Thanks @Nerivec !! Great feedback. Specially loved your perf test. I created one as well but it was so ugly I didn't thought about quoting it here. This is my first ts/js coding so I have lot to learn from your comments. I will work on fixes tomorrow. Thanks again for spending time on this.

File: lib/util/logger.ts (hunk 1 above)
Comment (tomer-w) [reply lib/util/logger.ts:194]: Fixed.

File: lib/util/logger.ts (hunk 2 above)
Comment (tomer-w) [reply lib/util/logger.ts:199]: ok. Removed.

**Comment by tomer-w** [comment]
I submitted all changes. Thanks again for the CR!

File: lib/mqtt.ts [hunk 3]
```diff
@@ -196,7 +196,7 @@ export default class MQTT {
         }
 
         if (!skipLog) {
-            logger.info(`MQTT publish: topic '${topic}', payload '${payload}'`, NS);
+            logger.info(() => `MQTT publish: topic '${topic}', payload '${payload}'`, NS);
```
Comment (Nerivec) [inline lib/mqtt.ts:199]: Why these changes? It's just strings, no `stringify`.

Comment (tomer-w) [reply lib/mqtt.ts:199]: The payload string gets really long and if log is disabled I don't see a reason for allocating memory and concatenating it all together.

**Comment by Nerivec** [comment]
z2m uses `stringify` from a package instead of `JSON.stringify` in several places. Might want to go through these too.

File: test/controller.test.js [hunk 4]
```diff
@@ -325,21 +326,47 @@ describe('Controller', () => {
...
+        }
+        const transport = new DummyTransport();
+        logger.setTransportsEnabled(true);
+        let spyTransport = jest.spyOn(transport, 'log').mockImplementation();
+        logger.addTransport(transport);
+
         await controller.start();
         logger.debug.mockClear();
         await MQTT.events.message('zigbee2mqtt/skip-this-topic', 'skipped');
-        expect(logger.debug).toHaveBeenCalledWith("Received MQTT message on 'zigbee2mqtt/skip-this-topic' with data 'skipped'", LOG_MQTT_NS);
```
Comment (Koenkk) [inline test/controller.test.js:338]: Instead of having to do this can't we modify `stub/logger.js` to evaluate the lambda?

Comment (Koenkk) [reply test/controller.test.js:338]: For this specific test case it's possible, but there are also some test cases on which this cannot be used. So having the `stub/logger.js` modified would be nice anyway.

Comment (tomer-w) [reply test/controller.test.js:338]: I did both:
1. Changed the code to spy on eventBus as it really seems more robust.
2. Changed the stub logger implementation so it will be easy to spy on the actual logger message. I added an example of this in the last line of the 'Handle mqtt message' test case so you can see that it is simple to do.

File: lib/mqtt.ts (hunk 3 above)
Comment (tomer-w) [reply lib/mqtt.ts:199]: Perf difference for
```
logger.debug(`Received MQTT message on '${topic}' with data '${buf.toString()}'`, 'z2m:mqtt');
```
tracelevel: error (so the debug trace will not happen)
```
┌─────────┬─────────────────┬─────────────┬───────────────────┬──────────┬──────────┐
│ (index) │ Task Name       │ ops/sec     │ Average Time (ns) │ Margin   │ Samples  │
├─────────┼─────────────────┼─────────────┼───────────────────┼──────────┼──────────┤
│ 0       │ 'plain log'     │ '3,054,102' │ 327.428450712782  │ '±0.25%' │ 15270513 │
│ 1       │ 'plain log new' │ '8,876,274' │ 112.6598795835417 │ '±0.09%' │ 44381372 │
└─────────┴─────────────────┴─────────────┴───────────────────┴──────────┴──────────┘

```
tracelevel: debug
```
┌─────────┬─────────────────┬─────────────┬────────────────────┬──────────┬──────────┐
│ (index) │ Task Name       │ ops/sec     │ Average Time (ns)  │ Margin   │ Samples  │
├─────────┼─────────────────┼─────────────┼────────────────────┼──────────┼──────────┤
│ 0       │ 'plain log'     │ '3,294,581' │ 303.52870101541646 │ '±0.37%' │ 16472907 │
│ 1       │ 'plain log new' │ '3,374,663' │ 296.32586530505574 │ '±0.30%' │ 16873317 │
└─────────┴─────────────────┴─────────────┴────────────────────┴──────────┴──────────┘
```
So I think the change is not relevant only for JSON.stringify
<details>
<summary>Code</summary>

```javascript

// https://github.com/tinylibs/tinybench
import {Bench, hrtimeNow} from 'tinybench';
import logger from './logger.js';

const bench = new Bench({time: 5000, warmupTime: 1000, now: hrtimeNow});

logger.init(); // set to 'info' with silenced Console transport
logger.setLevel('error');

const topic = "zigbee2mqtt/bridge/config"
const message = '"commit":"5b6f176e","coordinator":{"meta":{"maintrel":0,"majorrel":38,"minorrel":114,"product":0,"revision":"0x26720700","transportrev":0},"type":"ConBee2/RaspBee2"},"log_level":"debug","network":{"channel":11,"extendedPanID":"0xdddddddddddddddd","panID":6754},"permit_join":true,"version":"1.40.1-dev"}';
const buf: Buffer = Buffer.from(message);

bench.add('plain log', function() {
  logger.debug(`Received MQTT message on '${topic}' with data '${buf.toString()}'`, 'z2m:mqtt');
});

bench.add('plain log new', function() {
  logger.debug(() => `Received MQTT message on '${topic}' with data '${buf.toString()}'`, 'z2m:mqtt');
});


await bench.warmup(); // make results more reliable, ref: https://github.com/tinylibs/tinybench/pull/50
await bench.run();

console.table(bench.table());

await logger.end();

```
<details>

**Comment by tomer-w** [comment]
> z2m uses `stringify` from a package instead of `JSON.stringify` in several places. Might want to go through these too.

I saw that usage only as part of `MQTT.publish` so it can not be optimized out.

**Comment by tomer-w** [comment]
I think I resolved all your comments. Let me know if anything else is needed. Thanks guys!!

File: test/stub/logger.js [hunk 5]
```diff
@@ -5,20 +5,24 @@ let namespacedLevels = {};
...
-const callTransports = (level, message, namespace) => {
-    if (transportsEnabled) {
-        for (const transport of transports) {
-            transport.log({level, message, namespace}, () => {});
-        }
-    }
-};
 
 const mock = {
+    callTransports: jest.fn().mockImplementation((level, message, namespace) => {
```
Comment (Nerivec) [inline test/stub/logger.js:10]: Remove this function/prop and put the code directly in `log`? Spying on `logger.log` will feel more "natural".

File: lib/mqtt.ts (hunk 3 above)
Comment (tomer-w) [reply lib/mqtt.ts:200]: Indeed. Not that big difference but still worth the effort if I'm already on it.

File: test/stub/logger.js (hunk 5 above)
Comment (tomer-w) [reply test/stub/logger.js]: Fixed. Looks better now. thanks.

**Comment by tomer-w** [comment]
I found couple more usages I could swap. I think the commit is ready to merge. Thanks for your support.

**Comment by tomer-w** [comment]
OK, now it is really ready. Sorry for the mess.

File: lib/zigbee.ts [hunk 6]
```diff
@@ -125,8 +128,8 @@ export default class Zigbee {
         });
 
         logger.info(`zigbee-herdsman started (${startResult})`);
-        logger.info(`Coordinator firmware version: '${stringify(await this.getCoordinatorVersion())}'`);
-        logger.debug(`Zigbee network parameters: ${stringify(await this.herdsman.getNetworkParameters())}`);
+        logger.info(() => `Coordinator firmware version: '${stringify(this.getCoordinatorVersion())}'`);
```
Comment (Nerivec) [inline lib/zigbee.ts:131]: `await` went missing here and below.

Comment (tomer-w) [reply lib/zigbee.ts:131]: Indeed. It didnt work with await. Was there an easy way to keep it? I tested it and it worked without it.

Comment (tomer-w) [reply lib/zigbee.ts:131]: This is my first ts/js change, so I don't consider myself an expert about the implications of this removal. I will do whatever you recommend about this one.

Comment (Nerivec) [reply lib/zigbee.ts:131]: It works likely because the result is cached, so it's more or less sync despite the async wrapping function.
Better to remove the callback and leave a plain string I think, it's only used once on start anyway, near-zero impact on performance.

Comment (tomer-w) [reply lib/zigbee.ts]: I reverted this change. Ready to merge :).

**Comment by tomer-w** [comment]
Thank you!! Amazing project which is centerpiece for home automation.
